### PR TITLE
Add button in submission details page leading to the contest

### PR DESCRIPTION
**Pull request by msmner**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/382

### CODE REVIEW

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss [hunk 1]
```diff
@@ -2,6 +2,11 @@
 @use 'src/styles/components';
 @use 'src/styles/fonts';
 
+.backBtnClassName {
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss:5]: backBtn

**Review by shsimeonova — CHANGES_REQUESTED** [review]

**Review by mishogv — APPROVED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/common/url-types.ts [hunk 2]
```diff
@@ -7,6 +7,10 @@ interface IStartContestUrlParams {
     official: boolean;
 }
 
+interface IGetContestByProblemParams {
+    problemId: number;
```
Comment (mishogv) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/common/url-types.ts:11]: Remove approve

**Review by mishogv — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs [hunk 3]
```diff
@@ -28,6 +28,18 @@ public async Task<IActionResult> Register(int id, [FromQuery] bool official)
...
 
+    /// <summary>
+    /// Gets the contest that the problem is part of.
+    /// </summary>
+    /// <param name="problemId">The id of the problem that is used to find the contest.</param>
+    /// <returns>Model containing information about the contest.</returns>
+    [HttpGet("{problemId:int}")]
+    [ProducesResponseType(typeof(ContestServiceModel), Status200OK)]
+    public async Task<IActionResult> ByProblem(int problemId)
+    => await this.contestsBusinessService
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs:39]: format here

**Review by shsimeonova — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 4]
```diff
@@ -90,6 +123,16 @@ const SubmissionDetails = () => {
                   type={HeadingType.secondary}
                   className={styles.taskHeading}
                 >
+                    <span className={styles.lessThan}>
+                        &lt;
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:127]: Isn't there an icon for this?

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 5]
```diff
@@ -23,12 +25,43 @@ const SubmissionDetails = () => {
...
+
+    useEffect(() => {
+        const reload = sessionStorage.getItem('reload');
+        if (isNil(reload)) {
+            sessionStorage.reload = true;
+            window.location.reload();
+        } else {
+            sessionStorage.removeItem('reload');
+        }
+    }, []);
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:43]: 1. What is the purpose of this?
2. Don't use `sessionStorage` directly. Use `use-services`. There is a `CacheService` implementation that takes a type of storage. Currently, only `localStorageService` is provided, but a `sessionStorageService` can be added.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 6]
```diff
@@ -23,12 +25,43 @@ const SubmissionDetails = () => {
...
+
+    useEffect(() => {
+        if (isNil(currentSubmission)) {
+            return;
+        }
+
+        getContestByProblemId(currentSubmission.problem.id);
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:50]: Destructure.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 7]
```diff
@@ -23,12 +25,43 @@ const SubmissionDetails = () => {
...
+    }, [ currentSubmission, getContestByProblemId ]);
 
     const submissionTitle = useMemo(
         () => `Submission №${currentSubmission?.id}`,
         [ currentSubmission?.id ],
     );
 
+    const participationType = useMemo(
+        () => contest.canBeCompeted
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:59]: Destructure.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 8]
```diff
@@ -90,6 +123,16 @@ const SubmissionDetails = () => {
                   type={HeadingType.secondary}
                   className={styles.taskHeading}
                 >
+                    <span className={styles.lessThan}>
+                        &lt;
+                    </span>
+                    <LinkButton
+                      type={LinkButtonType.secondary}
+                      size={ButtonSize.small}
+                      to={getRegisterContestTypeUrl({ id: contest.id, participationType })}
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:132]: Value should be in `useMemo()`

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx [hunk 9]
```diff
@@ -36,13 +36,15 @@ interface IContestsContext {
...
         currentPage: number;
+        contest: IIndexContestsType;
     };
     actions: {
         reload: () => Promise<void>;
         clearFilters: () => void;
         clearSorts: () => void;
         toggleParam: (param: IFilter | ISort) => void;
         changePage: (pageNumber: number) => void;
+        getContestByProblemId: (problemId: number) => void;
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:47]: `loadContestByProblemId`. `get...` implies a result is returned, but this is not the case.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx [hunk 10]
```diff
@@ -212,6 +222,22 @@ const ContestsProvider = ({ children }: IContestsProviderProps) => {
         [ currentPage, filters, sortingTypes ],
     );
 
+    const getContestByProblemId = useCallback((problemId: number) => {
+        if (isNil(problemId)) {
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:226]: This check is repeated in `SubmissionDetails.tsx`. Remove it from there.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 11]
```diff
@@ -71,6 +71,18 @@ public async Task<RegisterUserForContestServiceModel> RegisterUserForContest(int
             return registerModel;
         }
 
+        public async Task<ContestServiceModel> GetContestByProblem(int problemId)
+        {
+           var contestServiceModel = await this.contestsData.GetByProblemId<ContestServiceModel>(problemId);
+           if (contestServiceModel == null)
+           {
+               throw new BusinessServiceException("Contest not found");
```
Comment (Minkov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:79]: Constant.

**Review by Minkov — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs (hunk 3 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs:39]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 5 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:43]: The purpose was to refresh the page once after loading, so that there is no bug with undefined values in the url. Since it was just a temporary solution, I removed it.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 6 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:50]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 7 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:59]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 8 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:132]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx (hunk 9 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:47]: Renamed.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx (hunk 10 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:226]: I could not find where this check is being done in SubmissionDetails.tsx. I am checking there on row 38 only if the currentSubmission isn't null. If I do not do that, then I get a nullability error when destructuring the problem id.

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs (hunk 11 above)
Comment (msmner) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:79]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 4 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:127]: There was an icon for right arrow, so I made one for left arrow as well and am using it.

**Review by mishogv — APPROVED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/guidelines/icons/LeftArrowIcon.tsx [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+import IconSize from './common/icon-sizes';
+import Icon, { IIconProps } from './Icon';
+
+type ILeftArrowIconProps = IIconProps
+
+const LeftArrowIcon = ({
+    className = '',
+    size = IconSize.Medium,
+    helperText = '',
+}:ILeftArrowIconProps) => (
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/guidelines/icons/LeftArrowIcon.tsx:13]: Is this file formatted correctly?

**Review by shsimeonova — CHANGES_REQUESTED** [review]

**Review by shsimeonova — APPROVED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/guidelines/icons/LeftArrowIcon.tsx (hunk 12 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/guidelines/icons/LeftArrowIcon.tsx:13]: I think it is : )

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx [hunk 13]
```diff
@@ -101,14 +106,19 @@ const ContestsProvider = ({ children }: IContestsProviderProps) => {
...
 
     const {
-        get,
-        data,
+        get: getContests,
+        data: contestsData,
     } = useHttp(getAllContestsUrl as UrlType, getAllContestsUrlParams);
 
+    const {
+        get: getContest,
```
Comment (Minkov) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:118]: `getContestByProblemId`

File: Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs [hunk 14]
```diff
@@ -25,6 +25,11 @@ public ContestsDataService(OjsDbContext db, IDatesService dates)
             : base(db)
             => this.dates = dates;
 
+        public Task<TServiceModel> GetByProblemId<TServiceModel>(int id)
+            => this.DbSet
+                .FirstOrDefaultAsync(c => c.ProblemGroups.Any(pg => pg.Problems.Any(p => p.Id == id)))
+                .Map<TServiceModel>();
```
Comment (Minkov) [inline Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs:31]: Do this as follows:
```
=> this.DbSet
  .Where(condition)
  .MapCollection<TServiceModel>()
  .FirstOrDefault();
```

This way, we use the power of AutoMapper and do the _mapping_ on the database, not in memory

**Review by Minkov — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 15]
```diff
@@ -26,12 +31,37 @@ const SubmissionDetails = () => {
...
+
+        loadContestByProblemId(id);
+    }, [ currentSubmission, loadContestByProblemId ]);
 
     const submissionTitle = useMemo(
         () => `Submission №${currentSubmission?.id}`,
         [ currentSubmission?.id ],
     );
 
+    const { canBeCompeted } = contest;
```
[the remark this answers is not in the document]
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:56]: Fixed.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx (hunk 13 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/hooks/use-contests.tsx:118]: Renamed.

File: Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs (hunk 14 above)
Comment (msmner) [reply Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs:31]: Fixed.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss [hunk 16]
```diff
@@ -14,15 +26,21 @@
   @extend %columnsThree;
 }
 
+.leftArrow {
+  bottom: -8px;
+  color: colors.$light-blue-color;
+  left: -13px;
```
[the remark this answers is not in the document]
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss:32]: Done.

File: Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs (hunk 14 above)
Comment (msmner) [reply Services/UI/OJS.Services.Ui.Data/Implementations/ContestsDataService.cs]: Done.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/common/app-url-types.tsx [hunk 17]
```diff
@@ -3,8 +3,8 @@
 import { ContestParticipationType } from './constants';
 
 interface IRegisterForContestTypeUrlParams {
-    id: number;
-    participationType: ContestParticipationType;
+    id?: number;
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/common/app-url-types.tsx:6]: Why is this change needed?

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs [hunk 18]
```diff
@@ -30,6 +30,18 @@ public async Task<IActionResult> Register(int id, [FromQuery] bool official)
...
             .ToOkResult();
 
+    /// <summary>
+    /// Gets the contest that the problem is part of.
+    /// </summary>
+    /// <param name="problemId">The id of the problem that is used to find the contest.</param>
+    /// <returns>Model containing information about the contest.</returns>
+    [HttpGet("{problemId:int}")]
+    [ProducesResponseType(typeof(ContestServiceModel), Status200OK)]
+    public async Task<IActionResult> ByProblem(int problemId)
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/ContestsController.cs:40]: GetByProblem

**Review by shsimeonova — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/common/app-url-types.tsx (hunk 17 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/common/app-url-types.tsx:6]: So that I avoid a cast in SubmissionDetails when getting the registerContestTypeUrl, but I revert it and just cast it now.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss [hunk 19]
```diff
@@ -1,28 +1,47 @@
...
 %navigation {
   align-content: center;
   flex: 3;
 }
 
-.taskHeading {
-  margin-bottom: 24px;
-  text-align: center;
+.retestButton {
+  margin: 0 0 0 24px;
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss:44]: Is it needed to use 0 values and 24px rather than margin-top (or whatever the 4th value is for)?

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 20]
```diff
@@ -26,12 +31,40 @@ const SubmissionDetails = () => {
...
+        loadContestByProblemId(id);
+    }, [ currentSubmission, loadContestByProblemId ]);
 
     const submissionTitle = useMemo(
         () => `Submission №${currentSubmission?.id}`,
         [ currentSubmission?.id ],
     );
 
+    const canBeCompeted = useMemo(
+        () => contest?.canBeCompeted,
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:57]: Is this needed to be in a useMemo?

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx [hunk 21]
```diff
@@ -117,7 +155,7 @@ const SubmissionDetails = () => {
         [ currentSubmission, canAccessAdministration ],
     );
 
-    if (isNil(currentSubmission)) {
+    if (isNil(currentSubmission) || isNil(contest)) {
```
Comment (shsimeonova) [inline Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:158]: Seems odd that the whole page wont be rendered if the contest is not loaded. Can you work around this and not render some elements maybe?

**Review by shsimeonova — CHANGES_REQUESTED** [review]

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss (hunk 19 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.module.scss:44]: I am not sure, but the retest button is not in the scope of my issue. It was merged from v2 into my branch.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 20 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:57]: If I try to get it like this: const { canBeCompeted } = contest; - it does not work, because I have to check first for nullability of the contest, so using useMemo was the easiest solution to have canBeCompeted as a constant.

File: Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx (hunk 21 above)
Comment (msmner) [reply Servers/UI/OJS.Servers.Ui/ClientApp/src/components/submissions/details/SubmissionDetails.tsx:158]: I removed this check there and now if the contest is null, then the button to the contest is just disabled. As soon as it is loaded, the button is enabled.

**Review by shsimeonova — DISMISSED** [review]

File: Services/UI/OJS.Services.Ui.Models/Contests/ContestProblemServiceModel.cs [hunk 22]
```diff
@@ -24,7 +24,7 @@ public class ContestProblemServiceModel : IMapExplicitly
 
         public int OrderBy { get; set; }
 
-        public int MaximumPoints { get; set; }
+        public short MaximumPoints { get; set; }
```
Comment (Minkov) [inline Services/UI/OJS.Services.Ui.Models/Contests/ContestProblemServiceModel.cs:27]: Wierd. Any reason for that?

File: Services/UI/OJS.Services.Ui.Models/Contests/ContestServiceModel.cs [hunk 23]
```diff
@@ -1,212 +1,210 @@
-﻿namespace OJS.Services.Ui.Models.Contests
+namespace OJS.Services.Ui.Models.Contests;
```
Comment (Minkov) [inline Services/UI/OJS.Services.Ui.Models/Contests/ContestServiceModel.cs:1]: Make changes like this in a different PR. As far as I can see, there are no changes to this file.

I can see the reason for making the changes, as they are needed for our code style, but they should be made clean.

**Review by Minkov — APPROVED** [review]

**Review by shsimeonova — DISMISSED** [review]

File: Services/UI/OJS.Services.Ui.Models/Contests/ContestProblemServiceModel.cs (hunk 22 above)
Comment (msmner) [reply Services/UI/OJS.Services.Ui.Models/Contests/ContestProblemServiceModel.cs:27]: To match the property in the Problem class, which is also short.

File: Services/UI/OJS.Services.Ui.Models/Contests/ContestServiceModel.cs (hunk 23 above)
Comment (msmner) [reply Services/UI/OJS.Services.Ui.Models/Contests/ContestServiceModel.cs:1]: Got it.

**Review by shsimeonova — APPROVED** [review]

**Review by PetarNeshkov — APPROVED** [review]
